File: src/lib/llm/__tests__/wordPieceTokenizer.test.ts
import * as fs from "node:fs"
import * as path from "node:path"
import {
	WordPieceTokenizer,
	CLS_ID,
	SEP_ID,
	PAD_ID,
	MAX_SEQ_LEN,
} from "../wordPieceTokenizer"

const VOCAB_PATH = path.resolve(
	__dirname,
	"../../../../android/app/src/main/assets/llm/minilm-l6-v2-vocab.txt",
)

let tokenizer: WordPieceTokenizer

beforeAll(() => {
	const vocabText = fs.readFileSync(VOCAB_PATH, "utf-8")
	tokenizer = WordPieceTokenizer.fromVocabText(vocabText)
})

describe("WordPieceTokenizer", () => {
	it("encodes 'hello world' with correct framing and ids", () => {
		const encoded = tokenizer.encode("hello world")
		expect(encoded.seqLen).toBe(MAX_SEQ_LEN)
		expect(encoded.inputIds[0]).toBe(CLS_ID)
		expect(encoded.inputIds[1]).toBe(7592) // hello
		expect(encoded.inputIds[2]).toBe(2088) // world
		expect(encoded.inputIds[3]).toBe(SEP_ID)
		expect(encoded.inputIds[4]).toBe(PAD_ID)
		expect(encoded.attentionMask[0]).toBe(1)
		expect(encoded.attentionMask[4]).toBe(0)
	})

	it("lowercases and splits punctuation", () => {
		const pieces = tokenizer.tokenize("Hello, World!")
		expect(pieces).toEqual(["hello", ",", "world", "!"])
	})

	it("applies WordPiece ## continuation to OOV tokens", () => {
		const pieces = tokenizer.tokenize("unaffordable")
		expect(pieces.length).toBeGreaterThan(1)
		for (let i = 1; i < pieces.length; i++) {
			expect(pieces[i].startsWith("##")).toBe(true)
		}
	})

	it("strips accents (NFD + drop combining marks)", () => {
		const pieces = tokenizer.tokenize("café")
		expect(pieces.join("").replace(/##/g, "")).toBe("cafe")
	})

	it("truncates to maxLen while preserving [CLS] and [SEP]", () => {
		const longText = "hello ".repeat(200)
		const encoded = tokenizer.encode(longText, 16)
		expect(encoded.seqLen).toBe(16)
		expect(encoded.inputIds[0]).toBe(CLS_ID)
		expect(encoded.inputIds[15]).toBe(SEP_ID)
	})
})

File: android/app/src/test/java/com/steve1316/uma_android_automation/llm/WordPieceTokenizerTest.kt
package com.steve1316.uma_android_automation.llm

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

/**
 * Unit tests for [WordPieceTokenizer] against the `bert-base-uncased` / `all-MiniLM-L6-v2` shared vocabulary.
 *
 * Expected ids are the reference output of HuggingFace `BertTokenizerFast("bert-base-uncased")` on the same inputs -
 * keeping Kotlin output in lockstep with it is what guarantees the ONNX model receives the exact tensors it was
 * trained on.
 */
@DisplayName("WordPieceTokenizer Tests")
class WordPieceTokenizerTest {
    private val tokenizer: WordPieceTokenizer =
        java.io.File("src/main/assets/llm/minilm-l6-v2-vocab.txt")
            .inputStream()
            .use { WordPieceTokenizer.fromVocabStream(it) }

    @Test
    @DisplayName("encodes 'hello world' with correct framing and ids")
    fun encodesHelloWorld() {
        val encoded = tokenizer.encode("hello world")
        assertEquals(WordPieceTokenizer.MAX_SEQ_LEN, encoded.seqLen)
        assertEquals(WordPieceTokenizer.CLS_ID.toLong(), encoded.inputIds[0])
        assertEquals(7592L, encoded.inputIds[1], "hello -> 7592")
        assertEquals(2088L, encoded.inputIds[2], "world -> 2088")
        assertEquals(WordPieceTokenizer.SEP_ID.toLong(), encoded.inputIds[3])
        assertEquals(WordPieceTokenizer.PAD_ID.toLong(), encoded.inputIds[4], "pad after [SEP]")
        assertEquals(1L, encoded.attentionMask[0])
        assertEquals(0L, encoded.attentionMask[4])
    }

    @Test
    @DisplayName("lowercases and splits punctuation")
    fun lowercasesAndSplitsPunctuation() {
        val pieces = tokenizer.tokenize("Hello, World!")
        assertEquals(listOf("hello", ",", "world", "!"), pieces)
    }

    @Test
    @DisplayName("applies WordPiece ## continuation to OOV tokens")
    fun wordpieceContinuation() {
        // "unaffordable" is not in vocab; expected split is "una" + "##ff" + "##ord" + "##able" (or similar).
        val pieces = tokenizer.tokenize("unaffordable")
        assertTrue(pieces.size > 1, "OOV word should split into multiple wordpieces")
        assertTrue(pieces.drop(1).all { it.startsWith("##") }, "continuation pieces must be ##-prefixed")
    }

    @Test
    @DisplayName("strips accents (NFD + drop combining marks)")
    fun stripsAccents() {
        val pieces = tokenizer.tokenize("café")
        assertEquals("cafe", pieces.joinToString("").replace("##", ""))
    }

    @Test
    @DisplayName("truncates to maxLen while preserving [CLS] and [SEP]")
    fun truncatesToMaxLen() {
        val longText = "hello ".repeat(200)
        val encoded = tokenizer.encode(longText, maxLen = 16)
        assertEquals(16, encoded.seqLen)
        assertEquals(WordPieceTokenizer.CLS_ID.toLong(), encoded.inputIds[0])
        assertEquals(WordPieceTokenizer.SEP_ID.toLong(), encoded.inputIds[15])
    }
}
